src/plugins/dashboard/public/application/utils/create_dashboard_container.tsx

```diff
@@ -0,0 +1,522 @@
+  return readonlyMode || userHasNoPermissions;
+};
+
+const getEmptyScreenProps = (
+  shouldShowEditHelp: boolean,
+  isEmptyInReadOnlyMode: boolean,
+  stateContainer: DashboardAppStateContainer,
+  container: DashboardContainer,
+  services: DashboardServices
+): DashboardEmptyScreenProps => {
+  const { embeddable, uiSettings, http, notifications, overlays, savedObjects } = services;
+  const emptyScreenProps: DashboardEmptyScreenProps = {
+    onLinkClick: () => {
+      if (shouldShowEditHelp) {
+        if (container && !isErrorEmbeddable(container)) {
+          openAddPanelFlyout({
+            embeddable: container,
+            getAllFactories: embeddable.getEmbeddableFactories,
+            getFactory: embeddable.getEmbeddableFactory,
+            notifications,
+            overlays,
+            SavedObjectFinder: getSavedObjectFinder(savedObjects, uiSettings),
+          });
+        }
+      } else {
+        stateContainer.transitions.set('viewMode', ViewMode.EDIT);
+      }
+    },
+    showLinkToVisualize: shouldShowEditHelp,
+    uiSettings,
+    http,
+  };
+  if (shouldShowEditHelp) {
+    emptyScreenProps.onVisualizeClick = async () => {
+      const type = 'visualization';
+      const factory = embeddable.getEmbeddableFactory(type);
+      if (!factory) {
+        throw new EmbeddableFactoryNotFoundError(type);
+      }
+      await factory.create({} as EmbeddableInput, container);
+    };
+  }
+  if (isEmptyInReadOnlyMode) {
+    emptyScreenProps.isReadonlyMode = true;
+  }
+  return emptyScreenProps;
+};
+
+export const renderEmpty = (
+  container: DashboardContainer,
+  appStateContainer: DashboardAppStateContainer,
+  services: DashboardServices
+) => {
+  const { dashboardConfig } = services;
+  const appStateData = appStateContainer.getState();
+  const shouldShowEditHelp = getShouldShowEditHelp(appStateData, dashboardConfig);
+  const shouldShowViewHelp = getShouldShowViewHelp(appStateData, dashboardConfig);
+  const isEmptyInReadOnlyMode = shouldShowUnauthorizedEmptyState(appStateData, services);
+  const isEmptyState = shouldShowEditHelp || shouldShowViewHelp || isEmptyInReadOnlyMode;
+  return isEmptyState ? (
+    <DashboardEmptyScreen
+      {...getEmptyScreenProps(
+        shouldShowEditHelp,
+        isEmptyInReadOnlyMode,
+        appStateContainer,
+        container,
+        services
+      )}
+    />
+  ) : null;
+};
+
+const setCurrentIndexPatterns = (dashboardContainer: DashboardContainer) => {
+  let panelIndexPatterns: IndexPattern[] = [];
+  dashboardContainer.getChildIds().forEach((id) => {
+    const embeddableInstance = dashboardContainer.getChild(id);
+    if (isErrorEmbeddable(embeddableInstance)) return;
+    const embeddableIndexPatterns = (embeddableInstance.getOutput() as any).indexPatterns;
+    if (!embeddableIndexPatterns) return;
+    panelIndexPatterns.push(...embeddableIndexPatterns);
+  });
+  panelIndexPatterns = uniqBy(panelIndexPatterns, 'id');
+  return panelIndexPatterns;
```

Comment:
```suggestion
  const panelIndexPatterns = new Map<string, IndexPattern>();
  dashboardContainer.getChildIds().forEach((id) => {
    const embeddableInstance = dashboardContainer.getChild(id);
    if (isErrorEmbeddable(embeddableInstance)) return;
    
    const embeddableIndexPatterns = (embeddableInstance.getOutput() as any).indexPatterns;
    if (!Array.isArray(embeddableIndexPatterns)) return;
    
    embeddableIndexPatterns.forEach((embeddableIndexPattern) => {
      if (!panelIndexPatterns.has(embeddableIndexPattern.id)
        panelIndexPatterns.set(embeddableIndexPattern.id, embeddableIndexPattern);
    });
  });

  return Array.from(panelIndexPatterns.values());
```

or if we trust that duplicate `id` values are always the same objects or we don't care which of the duplicates is included:

```suggestion
  const panelIndexPatterns = new Map<string, IndexPattern>(
    dashboardContainer.getChildIds().flatMap((id) => {
      const embeddableInstance = dashboardContainer.getChild(id);
      if (isErrorEmbeddable(embeddableInstance)) return [];
      
      const embeddableIndexPatterns = (embeddableInstance.getOutput() as any).indexPatterns;
      
      return embeddableIndexPatterns?.map((embeddableIndexPattern) => [embeddableIndexPattern.id, embeddableIndexPattern]) || [];
    })
  );

  return Array.from(panelIndexPatterns.values());
```

PS, I have not tested either of these in the code; the goal was to get rid of `uniqBy`.

Reply:
I have created a comment here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3365#issuecomment-1627593824.

This logic was ported over, the dashboards plugin is a pretty complex plugin since it's a container embeddable that is managing its state and its children states so previous attempts to clean up certain logic has caused edge case failures.